core/primitives/src/congestion_info.rs

```diff
@@ -269,10 +248,113 @@ impl CongestionInfo {
     }
 }
 
-impl CongestionInfoV1 {
+/// The extended congestion info contains the congestion info and extra
+/// information extracted from the block that is needed for congestion control.
+///
+/// It has simpler interface and it should be used instead of using the
+/// [`CongestionInfo`] directly.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct ExtendedCongestionInfo {
+    congestion_info: CongestionInfo,
+    missed_chunks_count: u64,
+}
+
+impl ExtendedCongestionInfo {
+    pub fn new(congestion_info: CongestionInfo, missed_chunks_count: u64) -> Self {
+        Self { congestion_info, missed_chunks_count }
+    }
+
+    pub fn congestion_info(self) -> CongestionInfo {
+        self.congestion_info
+    }
+
     /// How much gas another shard can send to us in the next block.
     pub fn outgoing_limit(&self, sender_shard: ShardId) -> Gas {
-        let congestion = self.congestion_level();
+        self.congestion_info.outgoing_limit(sender_shard, self.missed_chunks_count)
+    }
+
+    /// How much gas we accept for executing new transactions going to any
+    /// uncongested shards.
+    pub fn process_tx_limit(&self) -> Gas {
+        self.congestion_info.process_tx_limit()
+    }
+
+    /// Whether we can accept new transaction with the receiver set to this shard.
+    pub fn shard_accepts_transactions(&self) -> bool {
+        self.congestion_info.shard_accepts_transactions(self.missed_chunks_count)
+    }
```

Comment:
Actually it's only two methods that rely on missed_chunks_count. An small alternative to having ExtendedCongestionInfo and CongestionInfo the same interface would be to make the congestion_info field public and allow runtime to call the unaffected methods on it directly. It would be less code here but a bit more hassle in apply().

Reply:
Yeah, both ways seem fine. But I like the access through new methods, as you have done now, better.